README.md
[![NPM version](https://badge.fury.io/js/murmurhash3js.png)](http://badge.fury.io/js/murmurhash3js) [![Build Status](https://travis-ci.org/pid/murmurHash3js.png)](https://travis-ci.org/pid/murmurhash3js)

# MurmurHash3js
**A javascript implementation of [MurmurHash3](http://code.google.com/p/smhasher/source/browse/trunk/MurmurHash3.cpp?spec=svn145&r=144)'s hashing algorithms.**

# Installation
> npm install murmurhash3js

> bower install murmurhash3js

## Usage
```javascript
    // browser
        <script type="text/javascript" src="murmurhash3js.min.js"></script>
    // server
        var murmurHash3 = require("murmurhash3js");
```

```javascript
// Return a 32bit hash as a unsigned int:
> murmurHash3.x86.hash32("I will not buy this record, it is scratched.")
  2832214938

// Return a 128bit hash as a unsigned hex:
> murmurHash3.x86.hash128("I will not buy this tobacconist's, it is scratched.")
  "9b5b7ba2ef3f7866889adeaf00f3f98e"
> murmurHash3.x64.hash128("I will not buy this tobacconist's, it is scratched.")
  "d30654abbd8227e367d73523f0079673"

// Specify a seed (defaults to 0):
> murmurHash3.x86.hash32("My hovercraft is full of eels.", 25)
  2520298415

// Rebind murmurHash3:
> somethingCompletelyDifferent = murmurHash3.noConflict()
> murmurHash3
  undefined
> somethingCompletelyDifferent.version
  "3.0.0" // get version
```

## Authors

* [Karan Lyons](https://github.com/karanlyons/)
* [Sascha Droste](https://github.com/pid/)

## Changlog
 [CHANGLELOG.md](https://github.com/pid/murmurHash3js/blob/master/CHANGELOG.md)

## MIT License

Copyright © 2012-2014 Karan Lyons, Sascha Droste

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

The software is provided "as is", without warranty of any kind, express or implied, including but not limited to the warranties of merchantability, fitness for a particular purpose and non-infringement. In no event shall the authors or copyright holders be liable for any claim, damages or other liability, whether in an action of contract, tort or otherwise, arising from, out of or in connection with the software or the use or other dealings in the software.


CHANGELOG.md
Changelog
=========

- v3.0.0 [Bugfix](https://github.com/karanlyons/murmurHash3.js/pull/2) this version is equal to fork [Version 2.1.6](https://github.com/karanlyons/murmurHash3.js/blob/c1778f75792abef7bdd74bc85d2d4e1a3d25cfe9/murmurHash3.js )
- v2.1.6 removed testing for Node.js v0.8
- v2.1.5 udate devDependencies
- v2.1.4 add to bower.io; last version running with Node.js v0.8.x
- v2.1.3 update devDependencies
- v2.1.2 initial version; fork from https://github.com/karanlyons/murmurHash3.js ([Version 2.1.2](https://github.com/karanlyons/murmurHash3.js/blob/03bac51479581ab53e3b224ac474f4df69a89029/murmurHash3.js))


test/test-server.js
/* global describe,it,require */

describe('murmurHash3js', function() {
	'use strict';

	it('x86', function(done) {
		var murmurHash3 = require('../');

		murmurHash3.x86.hash32("I will not buy this record, it is scratched.")
			.should.be.equal(2832214938);
		murmurHash3.x86.hash128("I will not buy this tobacconist's, it is scratched.")
			.should.be.equal("ef3f78669b5b7ba200f3f98e889adeaf");
		murmurHash3.x86.hash32("My hovercraft is full of eels.", 0)
			.should.be.equal(2953494853);
		murmurHash3.x86.hash32("My hovercraft is full of eels.", 25)
			.should.be.equal(2520298415);
		murmurHash3.x86.hash32("My hovercraft is full of eels.", 128)
			.should.be.equal(2204470254);

		murmurHash3.x86.hash32("I will not buy this record, it is scratched.").should.be.equal(2832214938);
		murmurHash3.x86.hash32("").should.be.equal(0);
		murmurHash3.x86.hash32("0").should.be.equal(3530670207);
		murmurHash3.x86.hash32("01").should.be.equal(1642882560);
		murmurHash3.x86.hash32("012").should.be.equal(3966566284);
		murmurHash3.x86.hash32("0123").should.be.equal(3558446240);
		murmurHash3.x86.hash32("01234").should.be.equal(433070448);
		murmurHash3.x86.hash32("", 1).should.be.equal(1364076727);

		murmurHash3.x86.hash128("").should.be.equal("00000000000000000000000000000000");
		murmurHash3.x86.hash128("0").should.be.equal("a5eb34f80ab2409ea5eb34f8a5eb34f8");
		murmurHash3.x86.hash128("01").should.be.equal("674f3b210f87acb4674f3b21674f3b21");
		murmurHash3.x86.hash128("012").should.be.equal("4c13d78ecd94fea54c13d78e4c13d78e");
		murmurHash3.x86.hash128("0123").should.be.equal("485d3536dc378fea485d3536485d3536");
		murmurHash3.x86.hash128("01234").should.be.equal("7b3b211635c5b3ee00ae108800ae1088");
		murmurHash3.x86.hash128("012345").should.be.equal("6ce1944bdb26dc75f825536af825536a");
		murmurHash3.x86.hash128("0123456").should.be.equal("86d15c02b708d0a1495d053b495d053b");
		murmurHash3.x86.hash128("01234567").should.be.equal("92160402aa22bf8463b83c5e63b83c5e");
		murmurHash3.x86.hash128("012345678").should.be.equal("75d48126571b5f674ca675dcd0205c30");
		murmurHash3.x86.hash128("0123456789").should.be.equal("2e528b330017a61e57a11235a5443f20");
		murmurHash3.x86.hash128("0123456789a").should.be.equal("921f15e438a2ed0fa971884f2caa7f97");
		murmurHash3.x86.hash128("0123456789ab").should.be.equal("b6982a7ecfaa93f904d3d08f53412b5d");
		murmurHash3.x86.hash128("0123456789abc").should.be.equal("cc6d9d01c970af1dfc11eee3dd00c683");
		murmurHash3.x86.hash128("0123456789abcd").should.be.equal("c0a5114d6f34d20a63f51794ae0d83c5");
		murmurHash3.x86.hash128("0123456789abcde").should.be.equal("4d0818c03c76c46da78673faadd433da");
		murmurHash3.x86.hash128("0123456789abcdef").should.be.equal("36aed30afb7d4409572b3bfd48ad1d9b");
		murmurHash3.x86.hash128("", 1).should.be.equal("54d201b988c4adec54d201b954d201b9");

		done();
	});

	it('x64', function(done) {
		var murmurHash3 = require('../');

		murmurHash3.x64.hash128("I will not buy this record, it is scratched.")
			.should.be.equal("c382657f9a06c49d4a71fdc6d9b0d48f");
		murmurHash3.x64.hash128("I will not buy this tobacconist's, it is scratched.")
			.should.be.equal("d30654abbd8227e367d73523f0079673");
		murmurHash3.x64.hash128("My hovercraft is full of eels.", 0)
			.should.be.equal("03e5e14d358c16d1e5ae86df7ed5cfcb");
		murmurHash3.x64.hash128("My hovercraft is full of eels.", 25)
			.should.be.equal("e85cec5bbbe05ddefccbf1b933fff845");
		murmurHash3.x64.hash128("My hovercraft is full of eels.", 128)
			.should.be.equal("898223700c20009cf8b163b4519c7a35");

		murmurHash3.x64.hash128("I will not buy this tobacconist's, it is scratched.").should.be.equal("d30654abbd8227e367d73523f0079673");
		murmurHash3.x64.hash128("").should.be.equal("00000000000000000000000000000000");
		murmurHash3.x64.hash128("0").should.be.equal("2ac9debed546a3803a8de9e53c875e09");
		murmurHash3.x64.hash128("01").should.be.equal("649e4eaa7fc1708ee6945110230f2ad6");
		murmurHash3.x64.hash128("012").should.be.equal("ce68f60d7c353bdb00364cd5936bf18a");
		murmurHash3.x64.hash128("0123").should.be.equal("0f95757ce7f38254b4c67c9e6f12ab4b");
		murmurHash3.x64.hash128("01234").should.be.equal("0f04e459497f3fc1eccc6223a28dd613");
		murmurHash3.x64.hash128("012345").should.be.equal("88c0a92586be0a2781062d6137728244");
		murmurHash3.x64.hash128("0123456").should.be.equal("13eb9fb82606f7a6b4ebef492fdef34e");
		murmurHash3.x64.hash128("01234567").should.be.equal("8236039b7387354dc3369387d8964920");
		murmurHash3.x64.hash128("012345678").should.be.equal("4c1e87519fe738ba72a17af899d597f1");
		murmurHash3.x64.hash128("0123456789").should.be.equal("3f9652ac3effeb248027a17cf2990b07");
		murmurHash3.x64.hash128("0123456789a").should.be.equal("4bc3eacd29d386297cb2d9e797da9c92");
		murmurHash3.x64.hash128("0123456789ab").should.be.equal("66352b8cee9e3ca7a9edf0b381a8fc58");
		murmurHash3.x64.hash128("0123456789abc").should.be.equal("5eb2f8db4265931e801ce853e61d0ab7");
		murmurHash3.x64.hash128("0123456789abcd").should.be.equal("07a4a014dd59f71aaaf437854cd22231");
		murmurHash3.x64.hash128("0123456789abcde").should.be.equal("a62dd5f6c0bf23514fccf50c7c544cf0");
		murmurHash3.x64.hash128("0123456789abcdef").should.be.equal("4be06d94cf4ad1a787c35b5c63a708da");
		murmurHash3.x64.hash128("", 1).should.be.equal("4610abe56eff5cb551622daa78f83583");

		done();
	});
});

Gruntfile.js
module.exports = function(grunt) {
    'use strict';

    // Project configuration.
    grunt.initConfig({
        pkg: grunt.file.readJSON('package.json'),

        uglify: {
            options: {
                banner: '/* <%= pkg.name %> v<%= pkg.version %> (c) 2012-<%= grunt.template.today("yyyy") %> <%= pkg.author.name %> | Maintainer: <%= pkg.contributors[0].name %> <%= pkg.homepage %> */',
                max_line_length: 500
            },
            build: {
                src: '<%= buildSourceFile %>',
                dest: '<%= pkg.name %>.min.js'
            }
        },

        jshint: {
            options: {},
            src: ['Gruntfile.js', '<%= sourceFiles %>', '<%= jsonFiles %>']
        },

        watch: {
            files: ['Gruntfile.js', 'lib/**/*.js', 'test/**/*.js', 'package.json'],
            tasks: ['jshint', 'mocha', 'uglify']
        },

        bumpup: ['package.json', 'bower.json'],

        release: {
            options: {
                bump: false, //default: true
                tagName: 'v<%= version %>', //default: '<%= version %>'
                commitMessage: 'release v<%= version %>', //default: 'release <%= version %>'
                tagMessage: 'tagging version v<%= version %>' //default: 'Version <%= version %>'
            }
        },

        replace: {
            files: {
                src: '<%= buildSourceFile %>',
                overwrite: true,
                replacements: [{
                    from: /'version': '\d{1,2}\.\d{1,3}\.\d{1,4}',/g,
                    to: "'version': '<%= pkg.version %>',"
                }, {
                    from: /murmurHash3js.js v\d{1,2}\.\d{1,3}\.\d{1,3}/g,
                    to: "murmurHash3js.js v<%= pkg.version %>"
                }]
            },
            readme: {
                src: 'README.md',
                overwrite: true,
                replacements: [{
                    from: /"\d{1,2}\.\d{1,3}\.\d{1,4}" \/\/ get version/g,
                    to: "\"<%= pkg.version %>\" \/\/ get version"
                }]
            }
        },

        // files
        buildSourceFile: 'lib/murmurHash3js.js ',
        sourceFiles: 'lib/**/*.js ',
        testFiles: 'test/**/*.js ',
        jsonFiles: '*.json '
    });

    grunt.registerTask('mocha', 'run mocha', function() {
        var done = this.async();
        require('child_process')
            .exec('npm test', function(err, stdout) {
                grunt.log.write(stdout);
                done(err);
            });
    });

    grunt.event.on('watch ', function(action, filepath) {
        grunt.log.writeln(filepath + ' has ' + action);
    });

    grunt.loadNpmTasks('grunt-contrib-uglify');
    grunt.loadNpmTasks('grunt-contrib-jshint');
    grunt.loadNpmTasks('grunt-contrib-watch');
    grunt.loadNpmTasks('grunt-release');
    grunt.loadNpmTasks('grunt-bumpup');
    grunt.loadNpmTasks('grunt-text-replace');

    // Default task(s).
    grunt.registerTask('default', ['replace', 'jshint', 'mocha', 'uglify']);

};


lib/murmurHash3.js
/* jshint -W086: true */
// +----------------------------------------------------------------------+
// | murmurHash3.js v2.1.1 (http://github.com/karanlyons/murmurHash.js)   |
// | A javascript implementation of MurmurHash3's x86 hashing algorithms. |
// |----------------------------------------------------------------------|
// | Copyright (c) 2012, 2013 Karan Lyons                                       |
// | Freely distributable under the MIT license.                          |
// +----------------------------------------------------------------------+

;(function(root, undefined) {
	'use strict';

	// Create a local object that'll be exported or referenced globally.
	var library = {
		'version': '2.1.1', // change version only in package.json, will be replaced by grunt task
		'x86': {},
		'x64': {}
	};

	// PRIVATE FUNCTIONS
	// -----------------

	function _x86Multiply(m, n) {
		//
		// Given two 32bit ints, returns the two multiplied together as a
		// 32bit int.
		//

		return ((m & 0xffff) * n) + ((((m >>> 16) * n) & 0xffff) << 16);
	}

	function _x86Rotl(m, n) {
		//
		// Given a 32bit int and an int representing a number of bit positions,
		// returns the 32bit int rotated left by that number of positions.
		//

		return (m << n) | (m >>> (32 - n));
	}

	function _x86Fmix(h) {
		//
		// Given a block, returns murmurHash3's final x86 mix of that block.
		//

		h ^= h >>> 16;
		h = _x86Multiply(h, 0x85ebca6b);
		h ^= h >>> 13;
		h = _x86Multiply(h, 0xc2b2ae35);
		h ^= h >>> 16;

		return h;
	}

	function _x64Add(m, n) {
		//
		// Given two 64bit ints (as an array of two 32bit ints) returns the two
		// added together as a 64bit int (as an array of two 32bit ints).
		//

		m = [m[0] >>> 16, m[0] & 0xffff, m[1] >>> 16, m[1] & 0xffff];
		n = [n[0] >>> 16, n[0] & 0xffff, n[1] >>> 16, n[1] & 0xffff];
		var o = [0, 0, 0, 0];

		o[3] += m[3] + n[3];
		o[2] += o[3] >>> 16;
		o[3] &= 0xffff;

		o[2] += m[2] + n[2];
		o[1] += o[2] >>> 16;
		o[2] &= 0xffff;

		o[1] += m[1] + n[1];
		o[0] += o[1] >>> 16;
		o[1] &= 0xffff;

		o[0] += m[0] + n[0];
		o[0] &= 0xffff;

		return [(o[0] << 16) | o[1], (o[2] << 16) | o[3]];
	}

	function _x64Multiply(m, n) {
		//
		// Given two 64bit ints (as an array of two 32bit ints) returns the two
		// multiplied together as a 64bit int (as an array of two 32bit ints).
		//

		m = [m[0] >>> 16, m[0] & 0xffff, m[1] >>> 16, m[1] & 0xffff];
		n = [n[0] >>> 16, n[0] & 0xffff, n[1] >>> 16, n[1] & 0xffff];
		var o = [0, 0, 0, 0];

		o[3] += m[3] * n[3];
		o[2] += o[3] >>> 16;
		o[3] &= 0xffff;

		o[2] += m[2] * n[3];
		o[1] += o[2] >>> 16;
		o[2] &= 0xffff;

		o[2] += m[3] * n[2];
		o[1] += o[2] >>> 16;
		o[2] &= 0xffff;

		o[1] += m[1] * n[3];
		o[0] += o[1] >>> 16;
		o[1] &= 0xffff;

		o[1] += m[2] * n[2];
		o[0] += o[1] >>> 16;
		o[1] &= 0xffff;

		o[1] += m[3] * n[1];
		o[0] += o[1] >>> 16;
		o[1] &= 0xffff;

		o[0] += (m[0] * n[3]) + (m[1] * n[2]) + (m[2] * n[1]) + (m[3] * n[0]);
		o[0] &= 0xffff;

		return [(o[0] << 16) | o[1], (o[2] << 16) | o[3]];
	}

	function _x64Rotl(m, n) {
		//
		// Given a 64bit int (as an array of two 32bit ints) and an int
		// representing a number of bit positions, returns the 64bit int (as an
		// array of two 32bit ints) rotated left by that number of positions.
		//

		n %= 64;

		if (n === 32) {
			return [m[1], m[0]];
		} else if (n < 32) {
			return [(m[0] << n) | (m[1] >>> (32 - n)), (m[1] << n) | (m[0] >>> (32 - n))];
		} else {
			n -= 32;
			return [(m[1] << n) | (m[0] >>> (32 - n)), (m[0] << n) | (m[1] >>> (32 - n))];
		}
	}

	function _x64LeftShift(m, n) {
		//
		// Given a 64bit int (as an array of two 32bit ints) and an int
		// representing a number of bit positions, returns the 64bit int (as an
		// array of two 32bit ints) shifted left by that number of positions.
		//

		n %= 64;

		if (n === 0) {
			return m;
		} else if (n < 32) {
			return [(m[0] << n) | (m[1] >>> (32 - n)), m[1] << n];
		} else {
			return [m[1] << (n - 32), 0];
		}
	}

	function _x64Xor(m, n) {
		//
		// Given two 64bit ints (as an array of two 32bit ints) returns the two
		// xored together as a 64bit int (as an array of two 32bit ints).
		//

		return [m[0] ^ n[0], m[1] ^ n[1]];
	}

	function _x64Fmix(h) {
		//
		// Given a block, returns murmurHash3's final x64 mix of that block.
		// (`[0, h[0] >>> 1]` is a 33 bit unsigned right shift. This is the
		// only place where we need to right shift 64bit ints.)
		//

		h = _x64Xor(h, [0, h[0] >>> 1]);
		h = _x64Multiply(h, [0xff51afd7, 0xed558ccd]);
		h = _x64Xor(h, [0, h[0] >>> 1]);
		h = _x64Multiply(h, [0xc4ceb9fe, 0x1a85ec53]);
		h = _x64Xor(h, [0, h[0] >>> 1]);

		return h;
	}

	// PUBLIC FUNCTIONS
	// ----------------

	library.x86.hash32 = function(key, seed) {
		//
		// Given a string and an optional seed as an int, returns a 32 bit hash
		// using the x86 flavor of MurmurHash3, as an unsigned int.
		//

		key = key || '';
		seed = seed || 0;

		var remainder = key.length % 4;
		var bytes = key.length - remainder;

		var h1 = seed;

		var k1 = 0;

		var c1 = 0xcc9e2d51;
		var c2 = 0x1b873593;

		for (var i = 0; i < bytes; i = i + 4) {
			k1 = ((key.charCodeAt(i) & 0xff)) | ((key.charCodeAt(i + 1) & 0xff) << 8) | ((key.charCodeAt(i + 2) & 0xff) << 16) | ((key.charCodeAt(i + 3) & 0xff) << 24);

			k1 = _x86Multiply(k1, c1);
			k1 = _x86Rotl(k1, 15);
			k1 = _x86Multiply(k1, c2);

			h1 ^= k1;
			h1 = _x86Rotl(h1, 13);
			h1 = _x86Multiply(h1, 5) + 0xe6546b64;
		}

		k1 = 0;

		switch (remainder) {
			case 3:
				k1 ^= (key.charCodeAt(i + 2) & 0xff) << 16;

			case 2:
				k1 ^= (key.charCodeAt(i + 1) & 0xff) << 8;

			case 1:
				k1 ^= (key.charCodeAt(i) & 0xff);
				k1 = _x86Multiply(k1, c1);
				k1 = _x86Rotl(k1, 15);
				k1 = _x86Multiply(k1, c2);
				h1 ^= k1;
		}

		h1 ^= key.length;
		h1 = _x86Fmix(h1);

		return h1 >>> 0;
	};

	library.x86.hash128 = function(key, seed) {
		//
		// Given a string and an optional seed as an int, returns a 128 bit
		// hash using the x86 flavor of MurmurHash3, as an unsigned hex.
		//

		key = key || '';
		seed = seed || 0;

		var remainder = key.length % 16;
		var bytes = key.length - remainder;

		var h1 = seed;
		var h2 = seed;
		var h3 = seed;
		var h4 = seed;

		var k1 = 0;
		var k2 = 0;
		var k3 = 0;
		var k4 = 0;

		var c1 = 0x239b961b;
		var c2 = 0xab0e9789;
		var c3 = 0x38b34ae5;
		var c4 = 0xa1e38b93;

		for (var i = 0; i < bytes; i = i + 16) {
			k1 = ((key.charCodeAt(i) & 0xff)) | ((key.charCodeAt(i + 1) & 0xff) << 8) | ((key.charCodeAt(i + 2) & 0xff) << 16) | ((key.charCodeAt(i + 3) & 0xff) << 24);
			k2 = ((key.charCodeAt(i + 4) & 0xff)) | ((key.charCodeAt(i + 5) & 0xff) << 8) | ((key.charCodeAt(i + 6) & 0xff) << 16) | ((key.charCodeAt(i + 7) & 0xff) << 24);
			k3 = ((key.charCodeAt(i + 8) & 0xff)) | ((key.charCodeAt(i + 9) & 0xff) << 8) | ((key.charCodeAt(i + 10) & 0xff) << 16) | ((key.charCodeAt(i + 11) & 0xff) << 24);
			k4 = ((key.charCodeAt(i + 12) & 0xff)) | ((key.charCodeAt(i + 13) & 0xff) << 8) | ((key.charCodeAt(i + 14) & 0xff) << 16) | ((key.charCodeAt(i + 15) & 0xff) << 24);

			k1 = _x86Multiply(k1, c1);
			k1 = _x86Rotl(k1, 15);
			k1 = _x86Multiply(k1, c2);
			h1 ^= k1;

			h1 = _x86Rotl(h1, 19);
			h1 += h2;
			h1 = _x86Multiply(h1, 5) + 0x561ccd1b;

			k2 = _x86Multiply(k2, c2);
			k2 = _x86Rotl(k2, 16);
			k2 = _x86Multiply(k2, c3);
			h2 ^= k2;

			h2 = _x86Rotl(h2, 17);
			h2 += h3;
			h2 = _x86Multiply(h2, 5) + 0x0bcaa747;

			k3 = _x86Multiply(k3, c3);
			k3 = _x86Rotl(k3, 17);
			k3 = _x86Multiply(k3, c4);
			h3 ^= k3;

			h3 = _x86Rotl(h3, 15);
			h3 += h4;
			h3 = _x86Multiply(h3, 5) + 0x96cd1c35;

			k4 = _x86Multiply(k4, c4);
			k4 = _x86Rotl(k4, 18);
			k4 = _x86Multiply(k4, c1);
			h4 ^= k4;

			h4 = _x86Rotl(h4, 13);
			h4 += h1;
			h4 = _x86Multiply(h4, 5) + 0x32ac3b17;
		}

		k1 = 0;
		k2 = 0;
		k3 = 0;
		k4 = 0;

		switch (remainder) {
			case 15:
				k4 ^= key.charCodeAt(i + 14) << 16;

			case 14:
				k4 ^= key.charCodeAt(i + 13) << 8;

			case 13:
				k4 ^= key.charCodeAt(i + 12);
				k4 = _x86Multiply(k4, c4);
				k4 = _x86Rotl(k4, 18);
				k4 = _x86Multiply(k4, c1);
				h4 ^= k4;

			case 12:
				k3 ^= key.charCodeAt(i + 11) << 24;

			case 11:
				k3 ^= key.charCodeAt(i + 10) << 16;

			case 10:
				k3 ^= key.charCodeAt(i + 9) << 8;

			case 9:
				k3 ^= key.charCodeAt(i + 8);
				k3 = _x86Multiply(k3, c3);
				k3 = _x86Rotl(k3, 17);
				k3 = _x86Multiply(k3, c4);
				h3 ^= k3;

			case 8:
				k2 ^= key.charCodeAt(i + 7) << 24;

			case 7:
				k2 ^= key.charCodeAt(i + 6) << 16;

			case 6:
				k2 ^= key.charCodeAt(i + 5) << 8;

			case 5:
				k2 ^= key.charCodeAt(i + 4);
				k2 = _x86Multiply(k2, c2);
				k2 = _x86Rotl(k2, 16);
				k2 = _x86Multiply(k2, c3);
				h2 ^= k2;

			case 4:
				k1 ^= key.charCodeAt(i + 3) << 24;

			case 3:
				k1 ^= key.charCodeAt(i + 2) << 16;

			case 2:
				k1 ^= key.charCodeAt(i + 1) << 8;

			case 1:
				k1 ^= key.charCodeAt(i);
				k1 = _x86Multiply(k1, c1);
				k1 = _x86Rotl(k1, 15);
				k1 = _x86Multiply(k1, c2);
				h1 ^= k1;
		}

		h1 ^= key.length;
		h2 ^= key.length;
		h3 ^= key.length;
		h4 ^= key.length;

		h1 += h2;
		h1 += h3;
		h1 += h4;
		h2 += h1;
		h3 += h1;
		h4 += h1;

		h1 = _x86Fmix(h1);
		h2 = _x86Fmix(h2);
		h3 = _x86Fmix(h3);
		h4 = _x86Fmix(h4);

		h1 += h2;
		h1 += h3;
		h1 += h4;
		h2 += h1;
		h3 += h1;
		h4 += h1;

		return ("00000000" + (h2 >>> 0).toString(16)).slice(-8) + ("00000000" + (h1 >>> 0).toString(16)).slice(-8) + ("00000000" + (h4 >>> 0).toString(16)).slice(-8) + ("00000000" + (h3 >>> 0).toString(16)).slice(-8);
	};

	library.x64.hash128 = function(key, seed) {
		//
		// Given a string and an optional seed as an int, returns a 128 bit
		// hash using the x64 flavor of MurmurHash3, as an unsigned hex.
		//

		key = key || '';
		seed = seed || 0;

		var remainder = key.length % 16;
		var bytes = key.length - remainder;

		var h1 = [0, seed];
		var h2 = [0, seed];

		var k1 = [0, 0];
		var k2 = [0, 0];

		var c1 = [0x87c37b91, 0x114253d5];
		var c2 = [0x4cf5ad43, 0x2745937f];

		for (var i = 0; i < bytes; i = i + 16) {
			k1 = [((key.charCodeAt(i + 4) & 0xff)) | ((key.charCodeAt(i + 5) & 0xff) << 8) | ((key.charCodeAt(i + 6) & 0xff) << 16) | ((key.charCodeAt(i + 7) & 0xff) << 24), ((key.charCodeAt(i) & 0xff)) | ((key.charCodeAt(i + 1) & 0xff) << 8) | ((key.charCodeAt(i + 2) & 0xff) << 16) | ((key.charCodeAt(i + 3) & 0xff) << 24)];
			k2 = [((key.charCodeAt(i + 12) & 0xff)) | ((key.charCodeAt(i + 13) & 0xff) << 8) | ((key.charCodeAt(i + 14) & 0xff) << 16) | ((key.charCodeAt(i + 15) & 0xff) << 24), ((key.charCodeAt(i + 8) & 0xff)) | ((key.charCodeAt(i + 9) & 0xff) << 8) | ((key.charCodeAt(i + 10) & 0xff) << 16) | ((key.charCodeAt(i + 11) & 0xff) << 24)];

			k1 = _x64Multiply(k1, c1);
			k1 = _x64Rotl(k1, 31);
			k1 = _x64Multiply(k1, c2);
			h1 = _x64Xor(h1, k1);

			h1 = _x64Rotl(h1, 27);
			h1 = _x64Add(h1, h2);
			h1 = _x64Add(_x64Multiply(h1, [0, 5]), [0, 0x52dce729]);

			k2 = _x64Multiply(k2, c2);
			k2 = _x64Rotl(k2, 33);
			k2 = _x64Multiply(k2, c1);
			h2 = _x64Xor(h2, k2);

			h2 = _x64Rotl(h2, 31);
			h2 = _x64Add(h2, h1);
			h2 = _x64Add(_x64Multiply(h2, [0, 5]), [0, 0x38495ab5]);
		}

		k1 = [0, 0];
		k2 = [0, 0];

		switch (remainder) {
			case 15:
				k2 = _x64Xor(k2, _x64LeftShift([0, key.charCodeAt(i + 14)], 48));

			case 14:
				k2 = _x64Xor(k2, _x64LeftShift([0, key.charCodeAt(i + 13)], 40));

			case 13:
				k2 = _x64Xor(k2, _x64LeftShift([0, key.charCodeAt(i + 12)], 32));

			case 12:
				k2 = _x64Xor(k2, _x64LeftShift([0, key.charCodeAt(i + 11)], 24));

			case 11:
				k2 = _x64Xor(k2, _x64LeftShift([0, key.charCodeAt(i + 10)], 16));

			case 10:
				k2 = _x64Xor(k2, _x64LeftShift([0, key.charCodeAt(i + 9)], 8));

			case 9:
				k2 = _x64Xor(k2, [0, key.charCodeAt(i + 8)]);
				k2 = _x64Multiply(k2, c2);
				k2 = _x64Rotl(k2, 33);
				k2 = _x64Multiply(k2, c1);
				h2 = _x64Xor(h2, k2);

			case 8:
				k1 = _x64Xor(k1, _x64LeftShift([0, key.charCodeAt(i + 7)], 56));

			case 7:
				k1 = _x64Xor(k1, _x64LeftShift([0, key.charCodeAt(i + 6)], 48));

			case 6:
				k1 = _x64Xor(k1, _x64LeftShift([0, key.charCodeAt(i + 5)], 40));

			case 5:
				k1 = _x64Xor(k1, _x64LeftShift([0, key.charCodeAt(i + 4)], 32));

			case 4:
				k1 = _x64Xor(k1, _x64LeftShift([0, key.charCodeAt(i + 3)], 24));

			case 3:
				k1 = _x64Xor(k1, _x64LeftShift([0, key.charCodeAt(i + 2)], 16));

			case 2:
				k1 = _x64Xor(k1, _x64LeftShift([0, key.charCodeAt(i + 1)], 8));

			case 1:
				k1 = _x64Xor(k1, [0, key.charCodeAt(i)]);
				k1 = _x64Multiply(k1, c1);
				k1 = _x64Rotl(k1, 31);
				k1 = _x64Multiply(k1, c2);
				h1 = _x64Xor(h1, k1);
		}

		h1 = _x64Xor(h1, [0, key.length]);
		h2 = _x64Xor(h2, [0, key.length]);

		h1 = _x64Add(h1, h2);
		h2 = _x64Add(h2, h1);

		h1 = _x64Fmix(h1);
		h2 = _x64Fmix(h2);

		h1 = _x64Add(h1, h2);
		h2 = _x64Add(h2, h1);

		return ("00000000" + (h1[0] >>> 0).toString(16)).slice(-8) + ("00000000" + (h1[1] >>> 0).toString(16)).slice(-8) + ("00000000" + (h2[0] >>> 0).toString(16)).slice(-8) + ("00000000" + (h2[1] >>> 0).toString(16)).slice(-8);
	};

	// INITIALIZATION
	// --------------

	// Export murmurHash3 for CommonJS, either as an AMD module or just as part
	// of the global object.
	if (typeof exports !== 'undefined') {
		if (typeof module !== 'undefined' && module.exports) {
			exports = module.exports = library;
		}

		exports.murmurHash3 = library;
	} else if (typeof define === 'function' && define.amd) {
		define([], function() {
			return library;
		});
	} else {
		// Use murmurHash3.noConflict to restore `murmurHash3` back to its
		// original value. Returns a reference to the library object, to allow
		// it to be used under a different name.
		library._murmurHash3 = root.murmurHash3;

		library.noConflict = function() {
			root.murmurHash3 = library._murmurHash3;
			library._murmurHash3 = undefined;
			library.noConflict = undefined;

			return library;
		};

		root.murmurHash3 = library;
	}
})(this);

examples/node.js
var murmurHash3 = require("../");
// var murmurHash3 = require("murmurhash3js");

console.log("x86:");
console.log(murmurHash3.x86.hash32("I will not buy this record, it is scratched."));
console.log(murmurHash3.x86.hash128("I will not buy this tobacconist's, it is scratched."));
console.log(murmurHash3.x86.hash32("My hovercraft is full of eels.", 0));
console.log(murmurHash3.x86.hash32("My hovercraft is full of eels.", 25));
console.log(murmurHash3.x86.hash32("My hovercraft is full of eels.", 128));


console.log("\nx64:");
console.log(murmurHash3.x64.hash128("I will not buy this record, it is scratched."));
console.log(murmurHash3.x64.hash128("I will not buy this tobacconist's, it is scratched."));
console.log(murmurHash3.x64.hash128("My hovercraft is full of eels.", 0));
console.log(murmurHash3.x64.hash128("My hovercraft is full of eels.", 25));
console.log(murmurHash3.x64.hash128("My hovercraft is full of eels.", 128));
